pkg/generator/asset_generator.go

```diff
@@ -0,0 +1,363 @@
+			"${PORT_NAME}", port.Name,
+		}
+		var err error
+		serviceYAML, err = gen.applyAssetPatch(serviceYAML, "patches/metrics/service-port.yaml", extraReplacements)
+		if err != nil {
+			return err
+		}
+		serviceMonitorYAML, err = gen.applyAssetPatch(serviceMonitorYAML, "patches/metrics/service-monitor-port.yaml.patch", extraReplacements)
+		if err != nil {
+			return err
+		}
+	}
+
+	gen.generatedAssets.ControllerStaticResources[generated_assets.MetricServiceAssetName] = serviceYAML
+	if gen.flavour != FlavourHyperShift {
+		// TODO: figure out monitoring on HyperShift. The operator does not have RBAC for ServiceMonitors now.
+		gen.generatedAssets.ControllerStaticResources[generated_assets.MetricServiceMonitorAssetName] = serviceMonitorYAML
+	}
+	return nil
+}
+
+func (gen *AssetGenerator) collectControllerStaticAssets() error {
+	ctrlCfg := gen.operatorConfig.ControllerConfig
+	for _, a := range ctrlCfg.StaticAssets {
+		if a.ClusterFlavours.Has(gen.flavour) {
+			assetBytes := gen.mustReadAsset(a.AssetName, nil)
+			gen.generatedAssets.ControllerStaticResources[filepath.Base(a.AssetName)] = assetBytes
+		}
+	}
+	return nil
+}
+
+func (gen *AssetGenerator) generateGuest() error {
+	gen.generatedAssets.GuestStaticResources = make(map[string][]byte)
+	gen.generatedAssets.GuestStorageClassAssets = make(map[string][]byte)
+	gen.generatedAssets.GuestVolumeSnapshotClasses = make(map[string][]byte)
+
+	if err := gen.generateDaemonSet(); err != nil {
+		return err
+	}
+	if err := gen.collectGuestStaticAssets(); err != nil {
+		return err
+	}
+	if err := gen.collectGuestStorageClasses(); err != nil {
+		return err
+	}
+	if err := gen.collectGuestVolumeSnapshotClasses(); err != nil {
+		return err
+	}
+
+	if err := gen.patchGuest(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (gen *AssetGenerator) generateDaemonSet() error {
+	cfg := gen.operatorConfig.GuestConfig
+	dsYAML := gen.mustReadAsset("base/node.yaml", nil)
+	var err error
+
+	extraReplacements := []string{}
+	if cfg.LivenessProbePort > 0 {
+		extraReplacements = append(extraReplacements, "${LIVENESS_PROBE_PORT}", strconv.Itoa(int(cfg.LivenessProbePort)))
+	}
+
+	dsYAML, err = gen.applyAssetPatch(dsYAML, cfg.DaemonSetTemplateAssetName, extraReplacements)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(cfg.Sidecars); i++ {
+		sidecar := cfg.Sidecars[i]
+		dsYAML, err = gen.addSidecar(dsYAML, sidecar.TemplateAssetName, extraReplacements, sidecar.ExtraArguments, gen.flavour, sidecar.AssetPatches)
+		if err != nil {
+			return err
+		}
+	}
+	gen.generatedAssets.NodeTemplate = dsYAML
+	return nil
+}
+
+// Apply all patches in the generator config (CSIDriverGeneratorConfig.GuestConfig.AssetPatches)
+func (gen *AssetGenerator) patchGuest() error {
+	// Patch everything, including the CSI driver DaemonSet.
+	for _, patch := range gen.operatorConfig.GuestConfig.AssetPatches {
+		if !patch.ClusterFlavours.Has(gen.flavour) {
+			continue
+		}
+		switch patch.GeneratedAssetName {
```

Comment:
Same as for patchController(): let's have all mappings in a single map like `gen.generatedAssets.GuestAssests[]` and specific mappings like `gen.generatedAssets.GuestStorageClassAssets[]` are slices there. Btw, this switch is hard to review because `gen.operatorConfig.GuestConfig.AssetPatches` is empty. Was it intended?

Reply:
See above